app/main.py

```diff
@@ -86,129 +91,36 @@
 }
 
 
-def battle(knightsConfig):
-    # BATTLE PREPARATIONS:
-
-    # lancelot
-    lancelot = knightsConfig["lancelot"]
-
-    # apply armour
-    lancelot["protection"] = 0
-    for a in lancelot["armour"]:
-        lancelot["protection"] += a["protection"]
-
-    # apply weapon
-    lancelot["power"] += lancelot["weapon"]["power"]
-
-    # apply potion if exist
-    if lancelot["potion"] is not None:
-        if "power" in lancelot["potion"]["effect"]:
-            lancelot["power"] += lancelot["potion"]["effect"]["power"]
-
-        if "protection" in lancelot["potion"]["effect"]:
-            lancelot["protection"] += lancelot["potion"]["effect"]["protection"]
-
-        if "hp" in lancelot["potion"]["effect"]:
-            lancelot["hp"] += lancelot["potion"]["effect"]["hp"]
-
-    # arthur
-    arthur = knightsConfig["arthur"]
-
-    # apply armour
-    arthur["protection"] = 0
-    for a in arthur["armour"]:
-        arthur["protection"] += a["protection"]
-
-    # apply weapon
-    arthur["power"] += arthur["weapon"]["power"]
-
-    # apply potion if exist
-    if arthur["potion"] is not None:
-        if "power" in arthur["potion"]["effect"]:
-            arthur["power"] += arthur["potion"]["effect"]["power"]
-
-        if "protection" in arthur["potion"]["effect"]:
-            arthur["protection"] += arthur["potion"]["effect"]["protection"]
-
-        if "hp" in arthur["potion"]["effect"]:
-            arthur["hp"] += arthur["potion"]["effect"]["hp"]
-
-    # mordred
-    mordred = knightsConfig["mordred"]
-
-    # apply armour
-    mordred["protection"] = 0
-    for a in mordred["armour"]:
-        mordred["protection"] += a["protection"]
-
-    # apply weapon
-    mordred["power"] += mordred["weapon"]["power"]
-
-    # apply potion if exist
-    if mordred["potion"] is not None:
-        if "power" in mordred["potion"]["effect"]:
-            mordred["power"] += mordred["potion"]["effect"]["power"]
+def battle(knights: dict) -> dict:
+    dictionary_of_knights = {}
+    for knight_name, knight_data in knights.items():
+        knight_instance = Knight(
+            name=knight_data["name"],
+            power=knight_data["power"],
+            hp=knight_data["hp"],
+            armour=[Armour(armour) for armour in knight_data["armour"]],
+            weapon=Weapon(knight_data["weapon"]),
+            potion=Potion(knight_data["potion"]),
+        )
+        dictionary_of_knights[knight_name] = knight_instance
+
+    for knight in dictionary_of_knights:
+        dictionary_of_knights.get(knight).prepare_for_the_fight()
+        dictionary_of_knights.get(knight).apply_potion()
+
+    lancelot = dictionary_of_knights.get("lancelot")
+    mordred = dictionary_of_knights.get("mordred")
+    arthur = dictionary_of_knights.get("arthur")
+    red_knight = dictionary_of_knights.get("red_knight")
+
+    lancelot.fight_and_apply_damage(mordred)
```

Comment:
use loop instead of duplicating

Reply:
As in this particular task, each knight has its own rival that they fighting and not fighting every other knight. I believe the loop would not be much more efficient then hardcoding the battle between them.